File: multi_threading/python/performance_comparison.py
#!/usr/bin/env python3

# requires `requests`
# pip install requests

import requests
import time
import threading

URLS = [
    "https://www.{}".format(host)
    for host in (
        "wikipedia.org", "python.org", "fsf.org", "usaultimate.org", "npr.org",
        "reddit.com", "nytimes.com", "github.com", "scala-lang.org",
        "weather.gov", "apple.com", "whatsapp.com", "trello.com", "spotify.com",
        "duckduckgo.com", "mozilla.org", "canonical.com",
        "docker.com", "democracynow.org", "google.com",
    )
]

def get_and_print(url):
    response = requests.get(url)
    print("{}\t{}".format(response.status_code, response.url))


print("Making 20 HTTP GETs serially...")
start = time.time()
for url in URLS:
    get_and_print(url)

duration = time.time() - start
print("Serial GETs took {} seconds".format(duration))

print("\n---------\n")

print("Making 20 HTTP GETs in separate threads...")
start = time.time()
threads = []
for url in URLS:
    threads.append(
        threading.Thread(target=get_and_print, args=(url, ))
    )
for thread in threads:
    thread.start()
for thread in threads:
    thread.join()
duration = time.time() - start
print("Threaded GETs took {} seconds".format(duration))

# Sample output
# Making 20 HTTP GETs serially...
# 200 https://www.wikipedia.org/
# 200 https://www.python.org/
# ...
# 200 https://www.google.com/
# Serial GETs took 9.889546871185303 seconds
#
# ---------
#
# Making 20 HTTP GETs in separate threads...
# 200 https://www.apple.com/
# 200 https://www.wikipedia.org/
# ...
# 200 https://www.docker.com/
# Threaded GETs took 2.1607656478881836 seconds

File: multi_threading/go/performance_comparison.go
/**
 * Run with "go run performance_comparison.go"
 *
 * Requests 20 websites, in serial order and then goroutine'd.
 */
package main

import (
	"fmt"
	"net/http"
	"sync"
	"time"
)

func main() {
	urls := [20]string{
		"http://wikipedia.org", "http://python.org", "http://fsf.org", "http://usaultimate.org", "http://npr.org",
		"http://reddit.com", "http://nytimes.com", "http://github.com", "http://scala-lang.org",
		"http://weather.gov", "http://apple.com", "http://whatsapp.com", "http://trello.com", "http://spotify.com",
		"http://duckduckgo.com", "http://mozilla.org", "http://canonical.com",
		"http://docker.com", "http://democracynow.org", "http://google.com",
	}
	serial(urls)
	fmt.Println()
	parallel(urls)
}

func parallel(urls [20]string) {
	start := time.Now()
	var wg sync.WaitGroup
	wg.Add(20)
	for _, url := range urls {
		go getAndPrint(url, &wg)
	}
	wg.Wait()
	done := time.Since(start)
	fmt.Printf("Parallel took %s\n", done)
}

func serial(urls [20]string) {
	start := time.Now()
	for _, url := range urls {
		getAndPrint(url)
	}
	done := time.Since(start)
	fmt.Printf("Serial tool %s\n", done)
}

func getAndPrint(url string, wg ...*sync.WaitGroup) {
	response, err := http.Get(url)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(response.StatusCode, url)
	if len(wg) > 0 {
		defer wg[0].Done()
	}
}
